### PR TITLE
Switch to rust-lang-nursery/compiler-builtins

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit migrates the in-tree `libcompiler_builtins` to the upstream version
at https://github.com/rust-lang-nursery/compiler-builtins. The upstream version
has a number of intrinsics written in Rust and serves as an in-progress rewrite
of compiler-rt into Rust. Additionally it also contains all the existing
intrinsics defined in `libcompiler_builtins` for 128-bit integers.

It's been the intention since the beginning to make this transition but
previously it just lacked the manpower to get done. As this PR likely shows it
wasn't a trivial integration! Some highlight changes are:

* The PR rust-lang-nursery/compiler-builtins#166 contains a number of fixes
  across platforms and also some refactorings to make the intrinsics easier to
  read. The additional testing added there also fixed a number of integration
  issues when pulling the repository into this tree.

* LTO with the compiler-builtins crate was fixed to link in the entire crate
  after the LTO process as these intrinsics are excluded from LTO.

* Treatment of hidden symbols was updated as previously the
  `#![compiler_builtins]` crate would mark all symbol *imports* as hidden
  whereas it was only intended to mark *exports* as hidden.

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @japaric

**Comment by alexcrichton** [comment]
Note that some test updates and treatment of LTO are the same fixes as in https://github.com/rust-lang/rust/pull/42727, just pulled over here as well to get tests passing.

File: src/librustc_trans/trans_item.rs [hunk 1]
```diff
@@ -161,6 +161,18 @@ impl<'a, 'tcx> TransItem<'tcx> {
...
 
+        // If we're compiling the compiler-builtins crate, e.g. the equivalent of
+        // compiler-rt, then we want to implicitly compile everything with hidden
+        // visibility as we're going to link this object all over the place but
+        // don't want the symbols to get exported.
+        if linkage != llvm::Linkage::InternalLinkage &&
+           linkage != llvm::Linkage::PrivateLinkage &&
+           attr::contains_name(ccx.tcx().hir.krate_attrs(), "compiler_builtins") {
+            unsafe {
+                llvm::LLVMRustSetVisibility(lldecl, llvm::Visibility::Hidden);
```
Comment (nikomatsakis) [inline src/librustc_trans/trans_item.rs]: This seems probably like a better place, but I'm curious: why move this logic here from where it was before?

**Review by nikomatsakis — APPROVED** [review]
This looks pretty good to me, but I'm not intimately familiar with how this part of the system is all interconnected. Perhaps @eddyb, @arielb1, or @nagisa wants to take a look?

File: src/librustc_trans/trans_item.rs (hunk 1 above)
Comment (alexcrichton) [inline src/librustc_trans/trans_item.rs]: Ah an excellent question! Sorry this was a bit of a last minute change and forgot to mention it in the PR description. In the previous location this block would set *everything* with a `hidden` visibility, including function imports. What we wanted, however, is only our exported symbols to have `hidden` visibility, not the imports we had.

This manifest itself as failures when we tried to import `memcpy` as a hidden symbol from libc.so on ARM, which apparently failed. The intention here was to move this logic to where Rust functions are defined, not where every LLVM function is defined.

Is there a better place for this, though? I put it roughtly around where we seemed to extract function attributes and apply LLVM attributes, but I'd be fine placing this elsewhere.

File: src/bootstrap/bin/rustc.rs [hunk 2]
```diff
@@ -142,6 +142,11 @@ fn main() {
             }
         }
 
+        let crate_name = args.windows(2)
+            .find(|a| &*a[0] == "--crate-name")
+            .unwrap();
+        let crate_name = &*crate_name[1];
```
Comment (nagisa) [inline src/bootstrap/bin/rustc.rs]: Does this work if `--crate-name=foo` is used?

File: src/librustc_trans/trans_item.rs [hunk 3]
```diff
@@ -161,6 +161,18 @@ impl<'a, 'tcx> TransItem<'tcx> {
             llvm::SetUniqueComdat(ccx.llmod(), lldecl);
         }
 
+        // If we're compiling the compiler-builtins crate, e.g. the equivalent of
+        // compiler-rt, then we want to implicitly compile everything with hidden
+        // visibility as we're going to link this object all over the place but
+        // don't want the symbols to get exported.
+        if linkage != llvm::Linkage::InternalLinkage &&
+           linkage != llvm::Linkage::PrivateLinkage &&
+           attr::contains_name(ccx.tcx().hir.krate_attrs(), "compiler_builtins") {
```
Comment (nagisa) [inline src/librustc_trans/trans_item.rs]: Why not ccx.sess.cstore.is_compiler_builtins()?

File: src/test/run-pass/lib-defaults.rs [hunk 4]
```diff
@@ -8,16 +8,15 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// aux-build:clibrary.rs
```
Comment (nagisa) [inline src/test/run-pass/lib-defaults.rs]: The PR does not remove the clibrary.rs itself, does it?

File: src/bootstrap/bin/rustc.rs (hunk 2 above)
Comment (alexcrichton) [inline src/bootstrap/bin/rustc.rs]: No, but Cargo doesn't do that. This is only intended to ever work with Cargo.

File: src/librustc_trans/trans_item.rs (hunk 3 above)
Comment (alexcrichton) [inline src/librustc_trans/trans_item.rs]: Because the previous version didn't do that. I'll switch to using it.

Comment (alexcrichton) [inline src/librustc_trans/trans_item.rs]: Oh that's actually intended for foreign crates, not the local crate.

**Comment by Mark-Simulacrum** [comment]
This needs to also update the `init_repo` script as in https://github.com/rust-lang/rust/pull/42928.

**Comment by alexcrichton** [comment]
@aidanhs mind helping out with that? The failure mode here is that the `src/libcompiler_builtins` folder is switching from a folder to a git submodule, which I believe the `init_repo.sh` script doesn't handle just yet. Do you have a suggestion on the best way to fix this? Perhaps just a local workaround to blow away the `libcompiler_builtins` folder temporarily?

**Comment by aidanhs** [comment]
I'm not by a computer to try it out, but I think you should just be able to change `if [ ! -d "$cache_src_dir/$module" ]` to `if [ ! -e "$cache_src_dir/$module/.git" ]` - this will do a more robust check for a submodule, rather than assuming it's a git repo if the directory exists.

**Comment by aidanhs** [comment]
As an aside, it does make me a but sad that nested submodules are enabled unconditionally because it will now pointlessly clone the rust installer test repos.

File: src/librustc_trans/trans_item.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc_trans/trans_item.rs]: This seems like roughly the right spot to me.

**Comment by alexcrichton** [comment]
@aidanhs ok thanks! I'm now dubious this'll work on the bots, my guess is that the requirement for compiler-builtins to be recursive is going to break this...

**Comment by aidanhs** [comment]
@alexcrichton you did `-d` rather than `-e`, which won't work on some git versions.

You should also take the final two `--recursive` additions from https://github.com/rust-lang/rust/pull/42928/files#diff-4c597da51e91a8853ee60a8733858f71 (there's no point doing it when initialising submodules in the cache because they won't be used as a reference).

**Comment by aidanhs** [comment]
@bors r-

```
[00:02:39] tidy error: /checkout/src/ci/init_repo.sh:72: line longer than 100 chars
[00:02:40] some tidy checks failed
```

**Comment by kennytm** [comment]
Failed to build profiler on MSVC. Legit.

```
error: failed to run custom build command for `profiler_builtins v0.0.0 (file:///C:/projects/rust/src/libprofiler_builtins)`
process didn't exit successfully: `C:\projects\rust\build\x86_64-pc-windows-msvc\stage0-std\release\build\profiler_builtins-7f008c34aad7d625\build-script-build` (exit code: 101)
...
running: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\bin\\amd64\\cl.exe" "/nologo" "/MT" "/O2" "/Zl" "/Dstrdup=_strdup" "/Dopen=_open" "/Dfdopen=_fdopen" "/FoC:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage0-std\\x86_64-pc-windows-msvc\\release\\build\\profiler_builtins-bcaf90ff1dc76203\\out\\../libcompiler_builtins/compiler-rt/lib/profile\\WindowsMMap.o" "/c" "../libcompiler_builtins/compiler-rt/lib/profile\\WindowsMMap.c"
WindowsMMap.c
../libcompiler_builtins/compiler-rt/lib/profile\WindowsMMap.c(42): error C2065: 'DWORD': undeclared identifier
../libcompiler_builtins/compiler-rt/lib/profile\WindowsMMap.c(42): error C2146: syntax error: missing ';' before identifier 'flProtect'
../libcompiler_builtins/compiler-rt/lib/profile\WindowsMMap.c(42): error C2065: 'flProtect': undeclared identifier
../libcompiler_builtins/compiler-rt/lib/profile\WindowsMMap.c(45): error C2065: 'flProtect': undeclared identifier
...
```

**Comment by Mark-Simulacrum** [comment]
@bors retry

```
[00:49:27] thread 'main' panicked at 'client.read_exact(&mut header) failed with Connection reset by peer (os error 104)', /checkout/src/tools/remote-test-client/src/main.rs:240

```

**Comment by alexcrichton** [comment]
@bors: r-

Hm that looks suspicious, let me investigate that actually

**Comment by alexcrichton** [comment]
I believe the error is likely https://github.com/rust-lang-nursery/compiler-builtins/issues/173

**Comment by Mark-Simulacrum** [comment]
```
[00:52:51]           /checkout/src/liballoc_jemalloc/../jemalloc/include/jemalloc/internal/util.h:150: undefined reference to `__ffsdi2'

in all of

[00:52:51] failures:
[00:52:51]     [codegen-units] codegen-units/item-collection/cross-crate-generic-functions.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/cross-crate-trait-method.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/drop_in_place_intrinsic.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/function-as-argument.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/generic-drop-glue.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/generic-functions.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/generic-impl.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/impl-in-non-instantiated-generic.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/instantiation-through-vtable.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/items-within-generic-items.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/non-generic-drop-glue.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/non-generic-functions.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/static-init.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/statics-and-consts.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/trait-implementations.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/trait-method-as-argument.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/trait-method-default-impl.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/transitive-drop-glue.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/tuple-drop-glue.rs
[00:52:51]     [codegen-units] codegen-units/item-collection/unsizing.rs
[00:52:51]     [codegen-units] codegen-units/partitioning/vtable-through-const.rs
[00:52:51] 
```

**Comment by alexcrichton** [comment]
Linkers make no sense. How in the world did it get that far in the build before it found that error?

@bors: r=nikomatsakis

**Comment by est31** [comment]
Legit:
```
[00:47:59] ------------------------------------------
[00:47:59] stderr:
[00:47:59] ------------------------------------------
[00:47:59] error[E0432]: unresolved import `alloc_system::System`
[00:47:59]   --> /checkout/src/test/run-pass/smallest-hello-world.rs:22:5
[00:47:59]    |
[00:47:59] 22 | use alloc_system::System;
[00:47:59]    |     ^^^^^^^^^^^^^^^^^^^^ no `System` in the root
[00:47:59] 
[00:47:59] error: The attribute `global_allocator` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
[00:47:59]   --> /checkout/src/test/run-pass/smallest-hello-world.rs:24:1
[00:47:59]    |
[00:47:59] 24 | #[global_allocator]
[00:47:59]    | ^^^^^^^^^^^^^^^^^^^
[00:47:59]    |
[00:47:59]    = help: add #![feature(custom_attribute)] to the crate attributes to enable
[00:47:59] 
[00:47:59] error: aborting due to 2 previous errors
[00:47:59] 
[00:47:59] 
[00:47:59] ------------------------------------------
[00:47:59] 
[00:47:59] thread '[run-pass] run-pass/smallest-hello-world.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2473:8
```

File: src/test/run-pass/smallest-hello-world.rs [hunk 5]
```diff
@@ -14,10 +14,16 @@
...
 
 #![feature(intrinsics, lang_items, start, no_core, alloc_system)]
+#![feature(global_allocator, allocator_api)]
 #![no_core]
 
 extern crate alloc_system;
 
+use alloc_system::System;
+
+#[global_allocator]
```
Comment (est31) [inline src/test/run-pass/smallest-hello-world.rs]: Likely fallout from #42727

**Comment by est31** [comment]
Looks legit:

```
[01:28:22] Distcheck rust-src
[01:28:22] error: failed to load source for a dependency on `compiler_builtins`
[01:28:22] 
[01:28:22] Caused by:
[01:28:22]   Unable to update file:///checkout/obj/build/tmp/distcheck-src/rust-src/lib/rustlib/src/rust/src/rustc/compiler_builtins_shim
[01:28:22] 
[01:28:22] Caused by:
[01:28:22]   failed to read `/checkout/obj/build/tmp/distcheck-src/rust-src/lib/rustlib/src/rust/src/rustc/compiler_builtins_shim/Cargo.toml`
[01:28:22] 
[01:28:22] Caused by:
[01:28:22]   No such file or directory (os error 2)
[01:28:22] 
[01:28:22] 
[01:28:22] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "generate-lockfile" "--manifest-path" "/checkout/obj/build/tmp/distcheck-src/rust-src/lib/rustlib/src/rust/src/libstd/Cargo.toml"
[01:28:22] expected success, got: exit code: 101
[01:28:22] 
[01:28:22] 
[01:28:22] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test distcheck
[01:28:22] Build completed unsuccessfully in 1:26:59
```

**Comment by japaric** [comment]
@alexcrichton you are my hero. 🥇

**Pull request merged**